Added gates to Tiberian Sun

teees opened the pull request:
- Reuses some code from #7888
- Closes #3196
###### Works by
- Letting the pathfinder think there is no obstruction (for friendly and open gates)
- Blocking movement for closed gates and opens via OnNotifyBlockingMove
###### Improvements for the future:
- [fix zoffset for gates](https://github.com/OpenRA/OpenRA/pull/9981#issuecomment-174310942)
- placing gates on existing walls
- [fix the hitbox](https://github.com/OpenRA/OpenRA/pull/9981#issuecomment-170360964)

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+		public readonly string OpeningSound = "gateup1.aud";
+		public readonly string ClosingSound = "gatedwn1.aud";
+
+		[Desc("How many ticks until the gate will close after being opened.")]
```
"How many ticks until the gate starts closing after being opened?"

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+		[Desc("How many ticks until the gate will close after being opened.")]
+		public readonly int TicksToStayOpen = 50;
+
+		[Desc("How many ticks until the gate is open.")]
```
Is this the time it takes for the open/close animation to play? I don't understand the purpose from the name + description, so it probably needs a better description.

teees replied:
will change it, those are the ticks it takes to open the gate.
I made it independent of the actual animation, since ts has different animation lengths for nod/gdi. Gates are considered closed in transitions, that's why it's only used for opening

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public interface ITemporaryBlocker
```
Put this in `TraitInterfaces.cs`.

Phrohdoh replied:
Not all of our interfaces are there.
For example `IPreventsEjectOnDeath`.

It just depends on if it is specific to a trait, do we have another use for this currently?

teees replied:
I had it there before, when it was still called IGate but moved it since it's mod specific and isn't used in `OpenRA.Game`

penev92 replied:
`OpenRA.Mods.Common` has its own `TraitInterfaces.cs`.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+{
+	public interface ITemporaryBlocker
+	{
+		bool CanLiftBlockade(Actor self, Actor blocking);
```
`CanRemoveBlocker()` ?

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+		readonly WithSpriteBody wsb;
+
+		bool isOpen;
+		bool opening;
```
`isOpening`

penev92 replied:
Or perhaps change it to `isInTransition` and set it to true both while opening and closing?

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		int remainingTime;
+
+		public Gate(ActorInitializer init, GateInfo info) : base(init, info)
```
Put the base constructor on the next line to be consistent with the rest of the codebase please.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		void ITick.Tick(Actor self)
+		{
+			if (self.IsDisabled() || !BuildComplete)
```
Should probably consider `Locked` here as well, in case someone decides to capture them.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+			}
+		}
+
+		void Opened()
```
`OnOpened()` (or maybe `OnOpen()` ? ) please.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+				Opened();
+			else if (isOpen)
+			{
+				if (remainingTime-- <= 0)
```
Perhaps you should extract a single `remainingTime--` before the first check to avoid both confusion and bugs.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+					else
+						remainingTime = info.TicksToStayOpen;
+				}
+				else if (remainingTime % 10 == 0 && IsBlocked())
```
What's with the `%10` here?
This entire `if` seems redundant.

teees replied:
it checks for blocking units every 10 ticks. Checking every tick seemed unecessary to me

penev92 replied:
OK, but so does the first `if` - it checks for blockers every `remainingTime` ticks and even does the same when there are blockers.

pchote replied:
Just so we're clear, the idea here is to predict ahead of time that something is approaching so that it can start opening earlier, right?  The OnNotifyBlockingMove implementation will cover the case where it is actively blocking a unit.

teees replied:
No, currently the door will only open when "nudged" from OnNotifyBlockingMove. It is always considered to be in the way, since it's invisible to the pathfinder (for units that can actually open it). There is no prediction happening, I don't remember this being done in tiberian sun either. `IsBlocked()` checks if there is an actor standing on an opened gate, which will then wait for another `info.TicksToStayOpen` to close.

Enemies can still move into your base when they sneak on an opened gate as its influenced is removed from `ActorMap` while open

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		bool ITemporaryBlocker.CanLiftBlockade(Actor self, Actor blocking)
+		{
+			return !self.IsDisabled() && BuildComplete && blocking.AppearsFriendlyTo(self);
```
Again, you should check if `Locked` needs to be added here as well.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+		void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
+		{
+			if (!self.IsDisabled() && BuildComplete && !isOpen && (this as ITemporaryBlocker).CanLiftBlockade(self, blocking))
+				Open();
```
Same here regarding `Locked`.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
+		{
+			if (!self.IsDisabled() && BuildComplete && !isOpen && (this as ITemporaryBlocker).CanLiftBlockade(self, blocking))
```
Is it me, or are you basically doing `this.CanLiftBlocade(self, self)` here at the end?

teees replied:
thanks, changed

pchote commented:
Also don't forget about the animation stuff we talked about in IRC!

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+			foreach (var loc in eligibleLocations)
+			{
+				var blockers = self.World.ActorMap.GetActorsAt(loc).Where(a => a != self);
+				if (blockers != null && blockers.Any())
```
This can't be `null`.

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		bool IsBlocked()
+		{
+			var eligibleLocations = FootprintUtils.Tiles(self).ToList();
```
The `.ToList()` here should be unnecessary. You only enumerate it once in the `foreach`.

teees replied:
thanks, changed

penev92 commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,144 @@
+
+		void Close()
+		{
+			if (isOpen)
```
You probably want to make sure it's not currently opening/closing here.

pchote replied:
All these ugly flags and state is going to go away - it will be changed to work like the production doors instead

penev92 commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -571,6 +575,15 @@ public bool CanEnterCell(CPos cell, Actor ignoreActor = null, bool checkTransien
 			return Info.CanEnterCell(self.World, self, cell, ignoreActor, checkTransientActors ? CellConditions.All : CellConditions.BlockedByMovers);
 		}
 
+		public bool ContainsTemporaryBlocker(CPos cell)
```
This may fit better in `WorldUtils`. Just needs a `Actor ignoreActor = null` parameter like `CanEnterCell()`.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -721,3 +721,28 @@
+		HP: 350
+	Armor:
+		Type: Heavy
+	LineBuildNode:
```
This doesn't seem to work.  Walls are not visually connecting to the gates like they do for component towers.

Mailaender replied:
This seems to be resolved by now:

![image](https://cloud.githubusercontent.com/assets/756669/12215089/c7790370-b6ad-11e5-8559-9e2ababb4cd6.png)

pchote replied:
Yes.  I fixed that in e58ef101839f1d4ed4bccd3001ca7a6d011e7adc.

pchote commented:
I've pushed a prototype of what I was talking about earlier to e6f06e2fa21fc5dea26ba461658982fc06af714f.  This separates the world logic from rendering, removes most of the places where edge cases can hide in the state logic, and gives some hints on how to fix the wall connection issues.

abcdefg30 commented:
> OpenRA.Mods.Common/Traits/Buildings/Gate.cs:L15: [SA1505] An opening curly bracket must not be followed by a blank line.

pchote commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -271,6 +271,12 @@ public interface IMove
+	public interface ITemporaryBlocker
+	{
+		bool CanRemoveBlocker(Actor self, Actor blocking);
+		bool Blocks();
```
`bool Blocks(Actor self, CPos cell)` please.  Certain actor types may want to block only certain cells.

pchote commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -271,6 +271,12 @@ public interface IMove
 
+	public interface ITemporaryBlocker
+	{
+		bool CanRemoveBlocker(Actor self, Actor blocking);
```
Maybe `CanRemoveBlockage`?

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,15 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 
+		public static bool ContainsTemporaryBlocker(this World world, CPos cell, Actor ignoreActor = null)
+		{
+			var temporaryBlocker = world.ActorMap.GetActorsAt(cell).FirstOrDefault();
```
This should loop over all the actors in the cell.

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,15 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			var temporaryBlocker = world.ActorMap.GetActorsAt(cell).FirstOrDefault();
+			if (temporaryBlocker == null || temporaryBlocker == ignoreActor)
+				return false;
+			var temporaryBlockerTrait = temporaryBlocker.TraitOrDefault<ITemporaryBlocker>();
```
This should use `TraitsImplementing` and loop over all results.

teees commented:
@pchote I incorporated your changes, thank you

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,24 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			var temporaryBlockers = world.ActorMap.GetActorsAt(cell);
+			if (temporaryBlockers == null)
+				return false;
+			foreach (var temporaryBlocker in temporaryBlockers)
```
Empty line after the `return`, please.

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,24 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 			}
 		}
 
+		public static bool ContainsTemporaryBlocker(this World world, CPos cell, Actor ignoreActor = null)
```
You don't seem to be using `ignoreActor` at all.

teees replied:
fixed, thanks

penev92 commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,24 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+				var temporaryBlockerTraits = temporaryBlocker.TraitsImplementing<ITemporaryBlocker>();
+				if (temporaryBlockerTraits == null)
+					return false;
+				foreach (var temporaryBlockerTrait in temporaryBlockerTraits)
```
Empty line after the `return`, please.

penev92 commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -271,6 +271,12 @@ public interface IMove
+	public interface ITemporaryBlocker
+	{
+		bool CanRemoveBlockage(Actor self, Actor blocking);
+		bool Blocks(Actor self, CPos cell);
```
What happened to `IsBlocking()` ?

teees replied:
Might have been reverted when I merged it with @pchote 's branch, good catch

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+
+			foreach (var temporaryBlocker in temporaryBlockers)
+			{
+				if (temporaryBlocker == ignoreActor)
```
`ignoreActor` is used here.

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+			foreach (var temporaryBlocker in temporaryBlockers)
+			{
+				if (temporaryBlocker == ignoreActor)
+					return false;
```
This needs to be a continue, not a return.  There may be another blocking actor in the cell!

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+
+				var temporaryBlockerTraits = temporaryBlocker.TraitsImplementing<ITemporaryBlocker>();
+				if (temporaryBlockerTraits == null)
+					return false;
```
`continue` here too.

pchote commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+
+				foreach (var temporaryBlockerTrait in temporaryBlockerTraits)
+					if (temporaryBlockerTrait.IsBlocking(temporaryBlocker, cell))
+						return true;
```
This wants to stay as `return true;` since you know know that there definitely is a blocker.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,117 @@
+	public class GateInfo : BuildingInfo, ITraitInfo
+	{
+		public readonly string OpeningSound = "gateup1.aud";
+		public readonly string ClosingSound = "gatedwn1.aud";
```
Can we default these to `null` and then define them in yaml?

penev92 replied:
What's the standart? `null` or "" ? I know we're probably not very consistent with those, but still.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,117 @@
+				}
+			}
+
+			if (Position == OpenPosition)
```
braces around the body here

pchote commented on OpenRA.Mods.Common/Traits/Render/WithGateSpriteBody.cs:
```diff
@@ -0,0 +1,91 @@
+		}
+	}
+
+	// TODO: Add a new interface type for "things that walls can connect to.  It will have a method for AdjacentWallCanConnect
```
This was left as an exercise for you to finish :)

pchote commented on OpenRA.Mods.Common/Traits/Render/WithGateSpriteBody.cs:
```diff
@@ -0,0 +1,91 @@
+
+		static void UpdateNeighbours(Actor self)
+		{
+			// This code should stay basically the same, but requires an extra change in WithWallSpriteBody
```
this too

teees committed:
Added IWallConnector trait interface

teees committed:
Added Gates

FIXUP: account for full gate footprint when updating neighbours.

FIXUP: gate-wall connection adjacency yaml.

RoosterDragon commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+		public static bool ContainsTemporaryBlocker(this World world, CPos cell, Actor ignoreActor = null)
+		{
+			var temporaryBlockers = world.ActorMap.GetActorsAt(cell);
+			if (temporaryBlockers == null)
```
`GetActorsAt` will never return null, no need to check for it.

RoosterDragon commented on OpenRA.Game/WorldUtils.cs:
```diff
@@ -41,6 +41,29 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
+				if (temporaryBlocker == ignoreActor)
+					continue;
+
+				var temporaryBlockerTraits = temporaryBlocker.TraitsImplementing<ITemporaryBlocker>();
```
`TraitsImplementing` will never return null, no need to check for it.

teees replied:
thanks, fixed

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -251,6 +251,10 @@ bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, CellConditions
 
+			var temporaryBlocker = otherActor.TraitOrDefault<ITemporaryBlocker>();
+			if (temporaryBlocker != null && temporaryBlocker.CanRemoveBlockage(otherActor, self))
+				return false;
```
This seems like it should go above all the crush checks to me.

teees replied:
thanks, fixed

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Will open and be passable for actors that appear friendly when there are no enemies in range.")]
+	public class GateInfo : BuildingInfo, ITraitInfo
```
This extra `ITraitInfo` isn't needed (this is inherited from `BuildingInfo`).

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+			: base(init, info)
+		{
+			this.info = info;
+			this.self = init.Self;
```
You don't need the `this` here.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+		public readonly int TicksToStayOpen = 50;
+
+		[Desc("Ticks until the gate is considered open.")]
+		public readonly int TicksUntilOpen = 10;
```
Can we name this something like `TransitionDelay`?  It works in both directions, after all.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+		public readonly string ClosingSound = null;
+
+		[Desc("Ticks until the gate closes.")]
+		public readonly int TicksToStayOpen = 50;
```
This should also have `Delay` somewhere in the name (that's our convention for things that are waits measured in ticks)

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+
+		bool ITemporaryBlocker.IsBlocking(Actor self, CPos cell)
+		{
+			return Position != OpenPosition && FootprintUtils.Tiles(self).Contains(cell);
```
It would be worth implementing `INotifyAddedToWorld` and then caching `FootprintUtils.Tiles(self)`, because that does a bunch of work and this gets called a lot.

pchote commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,119 @@
+
+		bool IsBlocked()
+		{
+			var eligibleLocations = FootprintUtils.Tiles(self);
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithWallSpriteBody.cs:
```diff
@@ -97,8 +104,8 @@ public void Tick(Actor self)
 			{
-				var rb = a.TraitOrDefault<WithWallSpriteBody>();
-				if (rb == null || rb.wallInfo.Type != wallInfo.Type)
+				var rb = a.TraitOrDefault<IWallConnector>();
```
Style nit: can we rename this to `wc` to match the new type?

pchote commented on OpenRA.Mods.Common/Traits/Render/WithWallSpriteBody.cs:
```diff
@@ -97,8 +104,8 @@ public void Tick(Actor self)
-				var rb = a.TraitOrDefault<WithWallSpriteBody>();
-				if (rb == null || rb.wallInfo.Type != wallInfo.Type)
+				var rb = a.TraitOrDefault<IWallConnector>();
+				if (rb == null || !rb.AdjacentWallCanConnect(self, self.Location, wallInfo.Type))
```
This first arg needs to be `a` (since you're talking to a trait on the gate actor, not the wall).

pchote committed:
Add support for loading CVec[] from yaml.

pchote committed:
Fix wall adjacency for multi-cell neighbours.

pchote committed:
Fix line build for multi-cell neighbours.

pchote commented:
Here's some more commits to make this work better:
* Support loading `CVec[]` from yaml: 6708c4641d230b5ef03092ea61026d964ab70d3e
* Fix wall adjacency wrt multi-cell footprints: c783c1013c74e9138f58ff7b00fb1fc9f8d370ad
* Fix linebuild adjacency wrt multi-cell footprints: 98236b56366f10537d727b420b88d461f0bdd33a
* Fix gate adjacency wrt multi-cell footprints: a36dacb0b6d9a23175c29f78f596d71f22745545
* Yaml changes to set up the proper wall connections: 5d83ee1d990296bbdfc54ba12932d323af9e926f

These fix the last of the major issues I can find.

teees commented:
Will merge/rebase that fixup as soon as everything is sorted out

teees committed:
Add support for multiple IWallConnector traits

pchote commented on OpenRA.Mods.Common/Traits/Render/WithWallSpriteBody.cs:
```diff
@@ -97,20 +106,18 @@ public void Tick(Actor self)
-				var rb = a.TraitOrDefault<WithWallSpriteBody>();
-				if (rb == null || rb.wallInfo.Type != wallInfo.Type)
+				CVec facing;
+				var wc = a.TraitsImplementing<IWallConnector>().FirstOrDefault(b => b.IsTraitEnabled());
```
I think this should be `.FirstOrDefault(Exts.IsTraitEnabled);`?

atlimit8 replied:
This is an unnecessary lambda expression and the code usually uses `Where/FirstOrDefault/etc(Exts.IsTraitEnabled)`.
When I add the `Trait` base class, I plan on adding `public static bool IsEnabled` so that `Exts.IsTraitEnabled` is replaced by `Trait.IsEnabled`.

Mailaender commented on mods/ts/sequences/structures.yaml:
```diff
@@ -524,6 +524,84 @@ gawall:
+		Offset: -24, -24
+		UseTilesetCode: true
+	idle:
+		Length: 10
```
You can add `		ShadowStart: 21` here

GraionDilach replied:
`20`, not `21`.

teees replied:
21 should be ok I think, [0-9] is idle, [10-19] is damaged-idle, 20 is dead, [21-30] should then be idle shadow etc

Mailaender commented on mods/ts/sequences/structures.yaml:
```diff
@@ -524,6 +524,84 @@ gawall:
+		Length: 10
+	damaged-idle:
+		Start: 10
+		Length: 10
```
You can add `		ShadowStart: 21` here as well.

teees committed:
Add IBlocksProjectiles interface to avoid using BlocksProjectiles directly

teees commented:
I added an IBlocksProjectiles interface to allow linear scaling of blocksprojectiles height as well as disabling the trait (for firestorm gen)

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -237,6 +237,11 @@ bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, CellConditions
 				IsMovingInMyDirection(self, otherActor))
 				return false;
 
+			// If there is a closed gate / temporary blocker in our way and we can't open it, we are blocked.
```
Could this be changed to something like `We are not blocked by a temporary blocker that we can remove.` so you don't have to mentally negate the comment to match the code?

(I'm basically just asking this to be consistent with the other code & comments - parts that return false use "we are not blocked" whilst parts that return true use "we are blocked")

teees replied:
changed it, thanks

teees commented:
rebased

teees commented:
rebased once again

Phrohdoh commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -282,6 +282,12 @@ public interface IMove
 
+	public interface ITemporaryBlocker
+	{
+		bool CanRemoveBlockage(Actor self, Actor blocking);
```
This interface suggest the implementing trait (actor) is a blocker.
If so, why would they want to remove other blockage?

teees replied:
I guess you are referencing the `Actor self` argument? This was done to conform with the ongoing convention to have `self` as first argument. Not really any other reason.

Phrohdoh replied:
I mean having this method at all.

teees replied:
The interface suggests the actor is a _temporary_ blocker, `CanRemoveBlockage` will return different results for different `blocking` actors or when powered down etc. This is essentially how gates work, the pathfinder will use this function to see if the unit can potentially remove the blockage and if yes will just "bump" into it.

Phrohdoh replied:
This still makes no sense to me, but clearly I'm the only one, so ok.

pchote replied:
@Phrohdoh: `blocking` refers the actor that is blocked (i.e. the unit that wants to go through the gate), not another blocking actor actor that `self` may be able to remove.

Phrohdoh commented on OpenRA.Mods.Common/Traits/BlocksProjectiles.cs:
```diff
@@ -13,6 +13,11 @@
 {
+	public interface IBlocksProjectiles
+	{
+		WDist GetHeight(Actor self);
```
I think this could be named better, but this is probably in line with the current projectile code.

teees replied:
Traits would get the height directly from the traits info before, that's why it's just simply a getter now. Suggestions on a better name?

pchote replied:
Could become `WDist BlockingHeight { get; }`?

Phrohdoh commented:
![screen shot 2015-12-16 at 7 52 35 pm](https://cloud.githubusercontent.com/assets/4861023/11878747/388e5c98-a4bd-11e5-8cd1-20f1a64d55bb.png)

teees commented:
iirc @pchote mentioned that this will be resolved with proper z-buffer sorting

RoosterDragon commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -28,14 +30,14 @@ public class CombatDebugOverlay : IPostRender, INotifyDamage
 		{
 			healthInfo = self.Info.TraitInfoOrDefault<HealthInfo>();
-			blockInfo = self.Info.TraitInfoOrDefault<BlocksProjectilesInfo>();
+			blocker = self.TraitsImplementing<IBlocksProjectiles>().ToArray();
```
Any traits implementing `IBlocksProjectiles` are not guaranteed to have been created by the time you perform this lookup. Either wrap it in a `Lazy<T>` like the others do, or implement `INotifyCreated` and do your trait lookups in there (the traits will be ready by then).

teees commented:
rebased and adapted to @RoosterDragon's requested changes

teees commented:
and fixed my merge conflicts :/

Mailaender commented:
You can fix the sprite overlapping for the open mode
```
		ZOffset: -1c511
```
however that will conflict with the closed mode.

You may need a dynamic Z offset that adapts to the closed and open mode.

RoosterDragon commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -53,10 +59,11 @@ public void RenderAfterWorld(WorldRenderer wr, Actor self)
 
-			if (blockInfo != null)
+			var block = blocker.Where(Exts.IsTraitEnabled).ToList();
+			if (block.Any())
```
Maybe `blocker` -> `allBlockers` and `block` -> `blockers` for consistency with elsewhere in your PR where `blockers` are the enabled `IBlocksProjectiles` traits.

RoosterDragon commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -241,6 +241,11 @@ bool IsBlockedBy(Actor self, Actor otherActor, Actor ignoreActor, CellConditions
 				IsMovingInMyDirection(self, otherActor))
 				return false;
 
+			// If there is a temporary blocker in our path, but we can remove it, we are not blocked
```
Needs a full stop.

pchote commented:
> You may need a dynamic Z offset that adapts to the closed and open mode.

This has been discussed multiple times now, and

> iirc @pchote mentioned that this will be resolved with proper z-buffer sorting

teees commented:
rebased and fixed

pchote commented on mods/ts/rules/gdi-support.yaml:
```diff
@@ -25,6 +25,26 @@ GAWALL:
+		Prerequisites: gapile, ~structures.gdi
+	Tooltip:
+		Name: GDI Gate
+		Description: Stops infantry and blocks enemy fire.\nCan NOT be crushed by tanks.
```
I don't think anybody would expect gates to be crushable by tanks?  Seems like redundant info to me.

teees replied:
Oh you got me - I copy pasted that, probably from walls.

Phrohdoh replied:
Are walls even crushable in vanilla TS?

MustaphaTR replied:
Sandbags are but walls aren't.

pchote replied:
Kind of hard to come up with a description for something as simple as a gate... I suggest changing this to something like "Allows friendly units to pass through walls".

teees replied:
I changed it to "Automated barrier that opens up for allied units."

Mailaender commented:
It does block bullets shot diagonally when gates are open and shouldn't block anything:

![image](https://cloud.githubusercontent.com/assets/756669/12222140/a28c49f8-b7b3-11e5-9c86-4b1133d7cda7.png)

Mailaender commented:
Looks like bullets can pass closed gates when shot through a straight line:

![image](https://cloud.githubusercontent.com/assets/756669/12222184/0b32137e-b7b5-11e5-80db-69a39f3573b2.png)

MustaphaTR commented:
Bullets and other insta-hit projectiles have problems with blocking even on walls. It is still not fixed iirc.

pchote commented:
@MustaphaTR: that was fixed in #10243.

pchote commented:
Projectiles will be able to pass through the sides of the gate because we do not yet have a way to specify a proper hitbox for the gate (see #10349).  This doesn't explain why bullets are blocked by open gates, which is probably a bug.

teees committed:
Fix for BlockProjectiles to ignore blockers with height <= 0

teees commented:
Open gates shouldn't block anymore, the other issue is outside of the scope of this pr, as it is an already existing issue (see any line damage projectile missing (big) buildings if it doesn't traverse over the hitbox https://streamable.com/yoma )

Mailaender commented:
Works as promised. :white_check_mark: I consider the path-finding and wall integration very well done. Great work both @teees and @pchote :+1:

GraionDilach commented:
Can't the rectangle hitbox solve the footprint issue?

pchote commented:
Nope. The rectangle is in screen space.

pchote commented:
This already has my (so far implicit) :+1:, and nobody else seems to want to review it so I'll merge this tomorrow if no new review comments are made.

abcdefg30 commented on OpenRA.Game/FieldLoader.cs:
```diff
@@ -373,6 +373,28 @@ public static object GetValue(string fieldName, Type fieldType, MiniYaml yaml, M
+						return InvalidValueAction(value, fieldType, fieldName);
+
+					var vecs = new CVec[parts.Length / 2];
+					for (var i = 0; i < vecs.Length; ++i)
```
`i++`, please.

> (15:51:25) abcdefg30 [...] just wanted to know if we have a code style rule about that
(15:51:38) pchote nothing written down, but i++ is used more
(15:51:42) pchote so we should use that

abcdefg30 commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -220,12 +220,14 @@ void SanityCheckPath(Mobile mobile)
+
 			// Next cell in the move is blocked by another actor
-			if (!mobile.CanMoveFreelyInto(nextCell, ignoredActor, true))
+			if (!mobile.CanMoveFreelyInto(nextCell, ignoredActor, true) || containsTemporaryBlocker)
```
Would the order of the checks matter here? Like if `containsTemporaryBlockers` is already `true`, we'd save us the `CanMoveFreelyInto` check with this setup:
`if (containsTemporaryBlocker || !mobile.CanMoveFreelyInto(nextCell, ignoredActor, true))`?

teees replied:
The chance that the cell actually contains a temporary blocker is quite slim. Even in that case it wouldn't save you anything that much.

abcdefg30 commented on OpenRA.Mods.Common/Activities/Move/Move.cs:
```diff
@@ -220,12 +220,14 @@ void SanityCheckPath(Mobile mobile)
 				// Are we close enough?
 				var cellRange = nearEnough.Length / 1024;
-				if ((mobile.ToCell - destination.Value).LengthSquared <= cellRange * cellRange)
+				if ((mobile.ToCell - destination.Value).LengthSquared <= cellRange * cellRange && !containsTemporaryBlocker)
```
Similar question here.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:
```diff
@@ -72,12 +72,11 @@ public static IEnumerable<CPos> GetLineBuildCells(World world, CPos location, st
-						dirs[d] = -1; // Cell is blocked by another actor type
+					var hasConnector = world.ActorMap.GetActorsAt(cell)
+						.Any(a => a.Info.TraitInfos<LineBuildNodeInfo>()
+							.Any(info => info.Types.Overlaps(lbi.NodeTypes) && info.Connections.Contains(vecs[d])));
```
I don't think we want another level of indentation here, do we?

pchote replied:
It is nested inside the first `Any`, so the indentation is wanted.

abcdefg30 commented:
What is with https://github.com/OpenRA/OpenRA/pull/9981#discussion_r44780501?

teees commented:
It still derives from the Building class, so all the regular building locks should work as usual I'd guess. It would probably only matter if you want different behaviour for capturing closed and opened gates.

IMO gates count as base defense and shouldn't even be capturable - You are of course free to provide a patch for that in another pr.

TS gates weren't capturable https://github.com/LipkeGu/TSIniFiles/blob/master/rules.ini#L4895

RoosterDragon commented:
If another mod wishes to have capturable gates, we should make sure that works correctly (or at least make sure it doesn't make bad things happen and leave a TODO somewhere).

I'm not familiar enough with `Locked` to know what needs to happen (if anything) - @penev92 can you clarify?

teees commented:
Will add blocking then, probably tomorrow. Testing it however is tricky, see http://logs.openra.net/index.php?year=2016&month=01&day=17#13:35:12

teees commented:
rebased and added check for `Locked`

RoosterDragon commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,137 @@
+
+		void INotifyBlockingMove.OnNotifyBlockingMove(Actor self, Actor blocking)
+		{
+			if (Position != OpenPosition && (this as ITemporaryBlocker).CanRemoveBlockage(self, blocking))
```
`((ITemporaryBlocker)this)` please.

Side comment: Needing this cast is icky - a sad side effect of the explicit interfaces. (Don't change the interfaces or anything - I'm just mulling aloud here)

pchote replied:
IMO a better pattern would be to have a private implementation that is called from both places.

RoosterDragon replied:
Works for me.

teees replied:
fixed

RoosterDragon commented on OpenRA.Mods.Common/Traits/Buildings/Gate.cs:
```diff
@@ -0,0 +1,137 @@
+					return true;
+			}
+
+			return false;
```
There's an odd blend of imperative and functional style in this function.

I would choose one of these and stick with it:
```c#
return blockedPositions.Any(loc => self.World.ActorMap.GetActorsAt(loc).Any(a => a != self));
```

```c#
foreach (var loc in blockedPositions)
    foreach (var blocker in self.World.ActorMap.GetActorsAt(loc))
        if (blocker != self)
            return true;

return false;
```

teees replied:
fixed

RoosterDragon commented on OpenRA.Mods.Common/Traits/CombatDebugOverlay.cs:
```diff
@@ -53,10 +59,11 @@ public void RenderAfterWorld(WorldRenderer wr, Actor self)
 
-			if (blockInfo != null)
+			var blockers = allBlockers.Where(Exts.IsTraitEnabled).ToList();
+			if (blockers.Any())
```
Prefer `.Count > 0` to `.Any()`.

teees replied:
replaced

RoosterDragon commented on OpenRA.Mods.Common/Traits/Render/WithGateSpriteBody.cs:
```diff
@@ -0,0 +1,93 @@
+			var adjacent = OpenRA.Traits.Util.ExpandFootprint(footprint, true).Except(footprint)
+				.Where(self.World.Map.Contains).ToList();
+
+			var adjacentActorTraits = adjacent.SelectMany(c => self.World.ActorMap.GetActorsAt(c))
```
`.SelectMany(self.World.ActorMap.GetActorsAt)`

teees replied:
fixed

RoosterDragon commented:
There's a few loose ends of polish that have been picked up that we can clean up in the future (placement over walls, z-offset, blocking projectiles more correctly) but this is a very cool feature and the core mechanic of blocking and allowing movement seems to work just fine as far as I can see. I have a few last points of pedantry over the code style.

:+1: "Neat"

teees commented:
changed and rebased :)

RoosterDragon commented:
Needs another rebase.

Mailaender commented:
You forgot to account for the recent changes in https://github.com/OpenRA/OpenRA/pull/10537.

teees commented:
should be fixed again

pchote commented on OpenRA.Mods.Common/Traits/Render/WithWallSpriteBody.cs:
```diff
@@ -39,7 +40,7 @@ public override IEnumerable<IActorPreview> RenderPreviewSprites(ActorPreviewInit
 					foreach (var n in kv.Value)
 					{
-						var rb = init.World.Map.Rules.Actors[n].TraitInfoOrDefault<WithWallSpriteBodyInfo>();
+						var rb = init.World.Map.Rules.Actors[n].TraitInfos<WithWallSpriteBodyInfo>().FirstOrDefault(Exts.IsTraitEnabled);
```
Apologies for throwing up more roadblocks for getting this merged, but we're going to need an `IWallConnectorInfo` interface here to make sure that walls connect to gates in the map editor.

I'd also be happy to defer this to a followup PR.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -116,4 +116,10 @@ public interface IPreventsAutoTarget
 		bool PreventsAutoTarget(Actor self, Actor attacker);
 	}
+
+	interface IWallConnector
```
Could we get a `[RequireExplicitImplementation]` here?

pchote commented on OpenRA.Game/Traits/TraitsInterfaces.cs:
```diff
@@ -284,6 +284,12 @@ public interface IMove
 		bool CanEnterTargetNow(Actor self, Target target);
 	}
 
+	public interface ITemporaryBlocker
```
Could we get a `[RequireExplicitImplementation]` here too?  Now that we have that interface we may as well use it.

pchote commented on OpenRA.Mods.Common/Traits/BlocksProjectiles.cs:
```diff
@@ -12,6 +12,11 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	public interface IBlocksProjectiles
```
and a `[RequireExplicitImplementation]` here too if that doesn't require too many followup changes.

obrakmann commented:
Gates still have some z-ordering issues:
![screenshot from 2016-01-24 16 44 49](https://cloud.githubusercontent.com/assets/4331210/12537222/08a27eea-c2ba-11e5-881a-8c0dc55c2878.png)

However, I'm fine with merging this as-is and fixing the remaining issues in follow-ups. At least those should be merge-able faster than this one.

teees committed:
IWallConnectorInfo to allow gates connect to walls in map editor

teees commented:
Incorporated the requested changes

teees commented:
And rebased

Mailaender commented:
Still works, even after all those code reviews and rebases. Thanks for your endurance!

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/e623f0c014c33a384963396af475c96a3d0322a7...e1345029f195eddef5cc13df720becdcf77e3765)
